src/bun.js/bindings/napi.h

```diff
@@ -7,14 +7,144 @@
 #include "headers-handwritten.h"
 #include "BunClientData.h"
 #include <JavaScriptCore/CallFrame.h>
-
-#include "js_native_api.h"
+#include "node_api.h"
 #include <JavaScriptCore/JSWeakValue.h>
 #include "JSFFIFunction.h"
 #include "ZigGlobalObject.h"
 #include "napi_handle_scope.h"
 #include "napi_finalizer.h"
 
+#include <unordered_set>
+
+extern "C" void napi_internal_register_cleanup_zig(napi_env env);
+extern "C" void napi_internal_crash_in_gc(napi_env);
+
+struct napi_env__ {
+public:
+    napi_env__(Zig::GlobalObject* globalObject, const napi_module& napiModule)
+        : m_globalObject(globalObject)
+        , m_napiModule(napiModule)
+    {
+        napi_internal_register_cleanup_zig(this);
+    }
+
+    ~napi_env__()
+    {
+        delete[] filename;
+    }
+
+    void cleanup()
+    {
+        for (const BoundFinalizer& boundFinalizer : m_finalizers) {
+            Bun::NapiHandleScope handle_scope(m_globalObject);
+            if (boundFinalizer.callback) {
+                boundFinalizer.callback(this, boundFinalizer.hint, boundFinalizer.data);
+            }
+        }
+
+        m_finalizers.clear();
+    }
+
+    void removeFinalizer(napi_finalize callback, void* hint, void* data)
+    {
+        m_finalizers.erase({ callback, hint, data });
+    }
+
+    void addFinalizer(napi_finalize callback, void* hint, void* data)
+    {
+        m_finalizers.emplace(callback, hint, data);
+    }
+
+    void checkGC()
+    {
+        if (UNLIKELY(!mustAlwaysDefer() && m_globalObject->vm().heap.mutatorState() == JSC::MutatorState::Sweeping)) {
```

Comment:
Why do we check `mustAlwaysDefer()`? If finalizers are supposed to be deferred, then this check should still succeed, because we shouldn't be invoking the finalizer during GC in the first place.

Reply:
Good catch. I was missing an exception check too. There was a case where I was setting `*result` and returning `napi_ok` even with a RangeError pending.

Reply:
Seems I commented on the wrong feedback